Overhaul actor render and sprite bounds.

pchote opened the pull request:
The successive layers of workarounds and hacks in the render and selection bounds code recently passed their breaking point.  This PR aims to fix the underlying problem once and for all by throwing everything out and starting from a new, modern, base.

Most of the commits should be self explanatory and simple enough to review if you go through them one by one.  Things turn awful once we hit the the need to separate the selection bounds, and I couldn't find a cleaner way to separate out the commits without leaving broken code or regressions.  The pain of reviewing should be worth it (and minor compared to the pain of writing it).

Testing should use the new `/showscreenmap` command: the green rectangles are the areas used to work out when actors should be rendered on the screen, and the red rectangles mark the areas that are used for mouse interaction.  The visual selection boxes may be different again, as in the case of infantry. When I was testing I found it useful to move a unit's turret a stupid distance (several cells) away from the body to verify correct behaviour on turning and screen edge.

I expect there may be some minor regressions in selection box sizes (the algorithm for automatic sizing has changed), but these are best found by a fresh set of eyes, and can be easily fixed by defining overrides in yaml.  Third party modders will see loud errors/lint warnings if they have actors with too few or too many `IMouseBounds` traits.

Closes #4617
Fixes #5552
Fixes #6987
Fixes #11115
Fixes #12119
Fixes #14449
Fixes #14455
Fixes #14462
Fixes #14481
Fixes #14483
Fixes #14493
Supersedes #13926
Supersedes #14450
Implements most of #4783

pchote committed:
Simplify ScreenMap bounds checking.

pchote committed:
Use World.(AddTo|RemoveFrom)Maps in Immobile/Building.

pchote committed:
Add ISpriteSequence.Bounds and Animation(WithOffset).ScreenBounds.

The bounds define the smallest Rectangle that covers all frames within a sequence.

pchote committed:
Add IModel.AggregateBounds and ModelAnimation.ScreenBounds.

The bounds define the smallest Rectangle that covers all
rotations of all frames within a model sequence.

pchote committed:
Add ModelAnimation.IsVisible property.

pchote committed:
Update ScreenMap state in a single pass at the end of the tick.

pchote committed:
Add ScreenBounds method to IRender interface.

This method is expected to return Rectangles that cover
the region where Renderables returned by Render may exist.

pchote committed:
Add IActorPreview.ScreenBounds.

pchote committed:
Introduce IMouseBounds and split/rework mouse rectangles.

The render bounds for an actor now include the area covered
by bibs, shadows, and any other widgets. In many cases this
area is much larger than we really want to consider for
tooltips and mouse selection.

An optional Margin is added to Selectable to support cases
like infantry, where we want the mouse area of the actor
to be larger than the drawn selection box.

pchote committed:
Introduce IDecorationBounds to replace Actor.SelectionOverlayBounds.

pchote committed:
Select highest priority actor when not drag selecting.

pchote committed:
Increase infantry mouse bounds to improve usability.

pchote committed:
Add a lint test for conflicting Interactable and Selectable.

pchote committed:
Remove legacy bounds code.

reaperrr commented on OpenRA.Mods.Common/Traits/Render/WithTextDecoration.cs:
```diff
@@ -75,6 +75,12 @@ IEnumerable<IRenderable> IRender.Render(Actor self, WorldRenderer wr)
 			return !Info.RequiresSelection ? RenderInner(self, wr) : SpriteRenderable.None;
 		}
 
+		IEnumerable<Rectangle> IRender.ScreenBounds(Actor self, WorldRenderer wr)
```
Not sure, just a question (I already noticed this while working on my PR): Is using `IRender` instead of `IRenderAboveShroud` in this trait intentional?
Because it uses `IRenderAboveShroudWhenSelected` when `Info.RequiresSelection` is true, so this feels a bit inconsistent (note: I haven't checked yet whether a following commit adresses this).

pchote replied:
This was intentional because `IRenderAboveShroudWhenSelected` is invoked directly instead of going via the screen map.  #14082 complicates matters a bit, but i'd prefer to leave the discussion and implementation to that PR -- expanding the render bounds to include very large range circles doesn't feel like a good approach to me.

reaperrr commented on OpenRA.Game/Traits/Interactable.cs:
```diff
@@ -0,0 +1,80 @@
+		public readonly int[] Bounds = null;
+
+		[Desc("Defines a custom rectangle for Decorations (e.g. the selection box).",
+			"If null, CustomBounds will be used instead")]
```
Just 'Bounds' instead of 'CustomBounds'.

reaperrr commented on OpenRA.Mods.Common/Lint/CheckConflictingMouseBounds.cs:
```diff
@@ -0,0 +1,31 @@
+			{
+				var selectable = actorInfo.Value.TraitInfos<SelectableInfo>().Count();
+				var interactable = actorInfo.Value.TraitInfos<InteractableInfo>().Count();
+				if (selectable > 0 && selectable != interactable)
```
Won't this emit a warning even if `interactable == 0`?

pchote replied:
`SelectableInfo` inherits `InteractableInfo` so are also included in the `interactable` count.  The idea here is to make sure that there are either only `Interactable` traits (`selectable == 0`), or that there are no `Interactable` traits that are not also `Selectable` (`selectable == interactable`).

pchote commented:
Updated `[Desc]`.

pchote commented:
Note that this intentionally doesn't set up the map editor to use the new mouse interaction bounds.  That requires a whole new set of code that makes more sense as a followup PR considering that the bounds were already wrong there to start with.

reaperrr commented:
MiG shadow still disappears as soon as the mig itself is off-screen.

reaperrr commented:
The TD Landing Craft (not OLDLST, but the one with custom shp) and pretty much all trees and rocks in RA/TD have now too large interactable area, but IMO those can be dealt with in a follow-up.

Couldn't find any other regressions or problems. I'd like to see aircraft shadows fixed, but otherwise, this looks good to go to me. 👍

reaperrr commented:
Actually, I did come across a regression just now: The D2k harv "pick me up" symbol appeared somewhere in the middle of the shellmap.

pchote committed:
Add IRenderModifier.ModifyScreenBounds to support WithShadow.

pchote committed:
Use IRender.ScreenBounds in ScreenMap.

Traits are now required to trigger a ScreenMap update whenever they
believe that their ScreenBounds have changed.

pchote commented:
* Fixed aircraft shadows by adding <>
* Fixed selection decorations by removing `wr.ScreenPxPosition(self.CenterPosition)` (these are now baked into the bounds).

Smittytron commented:
Civilian buildings do not appear as frozen actors on explored map at the start of a game.

Also, it seemed like units lose their targets immediately as they enter the fog of war. (Maybe this is a good thing, I'll try to enlist a live body and see if I can assess how this might impact gameplay.)

pchote commented:
> Also, it seemed like units lose their targets immediately as they enter the fog of war.

This PR shouldn't be able to affect any of the targeting logic.  Can you please test the same cases on bleed, and try and track down the PR that introduced it?

pchote commented:
Fixed the frozen actor issue.  The problem was inside `FrozenUnderFog.TickRender`, where I was only updating the screen map inside the `if (renderables == null)`.  Fixed by caching the rectangles in the same was as the renderables and triggering an update for all dirty players.

abcdefg30 reviewed (approved):
@reaperrr is your 👍 still valid?

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/05866958643bd1f25f93142aa4b8eba3ea95b396...285c78e76269cb67bb09ef1fb5c5e56ad2a8f9dd
